pkg/gitnessclient/gitness_client.go

```diff
@@ -0,0 +1,289 @@
+	}
+
+	return &apiUser, nil
+}
+
+func (g *GitnessClient) GetRepositories(namespace string) ([]ApiRepository, error) {
+	space := ""
+	if namespace == personalNamespaceId {
+		user, err := g.GetUser()
+		if err != nil {
+			return nil, err
+		}
+		space = user.UID
+	} else {
+		space = namespace
+	}
+
+	reposURL, err := g.BaseURL.Parse(fmt.Sprintf("/api/v1/spaces/%s/+/repos", space))
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse url : %w", err)
+	}
+
+	body, err := g.performRequest("GET", reposURL.String())
+	if err != nil {
+		return nil, err
+	}
+
+	var apiRepos []ApiRepository
+	if err := json.Unmarshal(body, &apiRepos); err != nil {
+		return nil, err
+	}
+
+	return apiRepos, nil
+}
+
+func (g *GitnessClient) GetRepoBranches(repositoryId string, namespaceId string) ([]*apiRepoBranch, error) {
+	branchesURL, err := g.BaseURL.Parse(fmt.Sprintf("/api/v1/repos/%s/branches", url.PathEscape(namespaceId+"/"+repositoryId)))
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse url : %w", err)
+	}
+
+	body, err := g.performRequest("GET", branchesURL.String())
+	if err != nil {
+		return nil, err
+	}
+
+	var branches []*apiRepoBranch
+	if err := json.Unmarshal(body, &branches); err != nil {
+		return nil, fmt.Errorf("failed to unmarshal response body: %w", err)
+	}
+
+	return branches, nil
+}
+
+func (g *GitnessClient) GetRepoPRs(repositoryId string, namespaceId string) ([]*apiPR, error) {
+	prsURL, err := g.BaseURL.Parse(fmt.Sprintf("/api/v1/repos/%s/pullreq", url.PathEscape(namespaceId+"/"+repositoryId)))
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse url : %w", err)
+	}
+	values := url.Values{}
+	values.Add("state", "open")
+	apiUrl := prsURL.String() + "?" + values.Encode()
+
+	body, err := g.performRequest("GET", apiUrl)
+	if err != nil {
+		return nil, err
+	}
+
+	var apiPRs []*apiPR
+	if err := json.Unmarshal(body, &apiPRs); err != nil {
+		return nil, fmt.Errorf("failed to unmarshal response body: %w", err)
+	}
+
+	return apiPRs, nil
+}
+
+func (g *GitnessClient) GetLastCommitSha(repoURL string, branch *string) (string, error) {
+	ref := g.GetRepoRef(repoURL)
+
+	api, err := g.BaseURL.Parse(fmt.Sprintf("/api/v1/repos/%s/commits", url.PathEscape(ref)))
+	if err != nil {
+		return "", fmt.Errorf("failed to parse url : %w", err)
+	}
+
+	apiURL := ""
+	if branch != nil {
+		v := url.Values{}
+		v.Add("git_ref", *branch)
+		apiURL = api.String() + "?" + v.Encode()
+	} else {
+		apiURL = api.String()
+	}
+
+	body, err := g.performRequest("GET", apiURL)
+	if err != nil {
+		return "", fmt.Errorf("error while making request: %s", err.Error())
+	}
+
+	lastCommit, err := getLastCommit(body)
+	if err != nil {
+		return "", fmt.Errorf("error while fetching last commit from list: %s", err.Error())
+	}
+
+	return lastCommit.Sha, nil
+}
+
+func (g *GitnessClient) GetRepoRef(url string) string {
+	repoUrl := strings.TrimSuffix(url, ".git")
+	parts := strings.Split(repoUrl, "/")
```

Comment:
Server panics on `dtn create http://localhost:3000/Project` - I believe it's because there is no checks to the length of the `parts` slice - please add some checks and see if that resolves the panic

Also, the methods that call `GetRepoRef` might need some len checks before passing in the function parameter

Reply:
@idagelic  I am quite confused about lens check before passing in the function parameter. I was wondering if you could help me understand how we can perform a lens check before passing a value as a parameter to a function.

Reply:
@Rutik7066 sorry for bad wording on that part.

I just checked again and what I was referring to was just this instance:

![image](https://github.com/daytonaio/daytona/assets/25279767/da231f4c-e30a-4a7a-a1cc-ffcdb1b02ce0)

We should check if the len of prUrl is more than 1 before assigning the repo.Name as prUrl[1] might cause a panic as well when there is just one element and not two

Reply:
Sure I will add lens check rest of the changes is applied I commit the changes in minutes

Reply:
@idagelic  Please check the changes.

Reply:
Change this line
`if len(parts) < 2 {`
to
`if len(parts) < 5 {`

And add a `len(parts) >= 6` check directly above this line:
`path = fmt.Sprintf("%s/%s", parts[4], parts[5])`

otherwise the function may panic.